File: java/flesch.java
import java.util.*;
import java.io.*;

public class flesch {
  public static void main(String[] args)
  {
    if(args.length != 1)
    {
      System.err.println("Error: There must be exactly one command line argument.");
      System.exit(1);
    }

    ArrayList<String> words = getInput(args[0]);

    int numWords = words.size();
    int numSyllables = countTotalSyllables(words);
    int numSentences = countSentences(words);
    int numDifficultWords = countDifficultWords(words);

    //for debugging
    System.out.println("numWords = " + numWords);
    System.out.println("numSyllables = " + numSyllables);
    System.out.println("numSentences = " + numSentences);
    System.out.println("numDifficultWords = " + numDifficultWords);

    //Calculate Flesch Readability Index
    double alpha = (double) numSyllables / numWords;
    double beta = (double) numWords / numSentences;

    int fleschIndex = (int)Math.round(206.835 - alpha * 84.6 - beta * 1.015);

    //Calculate Flesch-Kincaid Grade Level Index
    double fleschKincaidIndex = alpha * 11.8 + beta * 0.39 - 15.59;
    fleschKincaidIndex = Math.round(fleschKincaidIndex * 10) / 10.0; //round to one decimal point

    //Calculate Dale-Chall Readability Score
    alpha = (double) numDifficultWords / numWords;
    double dalechall = alpha * 100 * 0.1579 + beta * 0.0496;
    if (alpha > 0.05)
      dalechall += 3.6365;
    dalechall = Math.round(dalechall *10) / 10.0; //round to one decimal point

    System.out.println("Flesch Readability Index = " + fleschIndex);
    System.out.println("Flesch-Kincaid Grade Level Index = " + fleschKincaidIndex);
    System.out.println("Dale-Chall Readability Score = " + dalechall);

  }

  private static ArrayList<String> getInput(String filename)
  {
    //returns an ArrayList of all the words in the input text file
    //all the valid filenames for the translations: ASV.txt  BBE.txt  DARBY.txt  DRB.txt  KJ21.txt  KJV.txt  NAB.txt  NASB.txt  NIV.txt  NKJV.txt  NLT.txt  WEB.txt  YLT.txt

    ArrayList<String> words = new ArrayList<String>();
    try
    {
      Scanner scan = new Scanner(new File("/pub/pounds/CSC330/translations/" + filename));
      int linesProcessed = 0;
      while (scan.hasNextLine())
      {
        String line = scan.nextLine();
        linesProcessed++;
        for(String token : line.split(" "))
        {
          if(isWord(token))
          {
            words.add(token);
          }
        }
      }
      scan.close();
      System.out.println("Last word processed is: " + words.get(words.size()-1));
      System.out.println("Lines Processed: " + linesProcessed);
    }
    catch (FileNotFoundException e)
    {
      System.err.println("Error: Specified file does not exist.");
      System.exit(1);
    }
    return words;
  }

  private static Boolean isWord(String s)
  {
    //returns true if there is at least one alphabetic character
    for(char c : s.toCharArray())
    {
      if(Character.isLetter(c))
        return true;
    }
    return false;
  }

  private static int countSentences(ArrayList<String> words)
  {
    int count = 0;
    for(String word: words)
    {
      char c = word.charAt(word.length()-1); //last character in word
      if (c == '.' || c == ':' || c == ';' || c == '?' || c == '!') //check if the last character is punctuation
      {
        count++;
      }
    }
    return count;
  }

  private static int countTotalSyllables(ArrayList<String> words)
  {
    int count = 0;
    for(String word: words)
    {
      count += countSyllables(word);
    }
    return count;
  }

  private static int countSyllables(String word)
  {
    word = stripWord(word);
    int numSyllables = countSyllablesRecursive(word);

    if (numSyllables == 0)
      numSyllables = 1; //no word can have 0 syllables

    //System.out.println(word + " " + numSyllables); //debugging TODO remove
    return numSyllables;
  }

  private static int countSyllablesRecursive(String word)
  {
    if(word.length() == 0) //empty string has no syllables
      return 0;

    if(isVowel(word.charAt(0)))
    {
      if(word.length() == 1) //if the word is only 1 character, then it is only 1 syllable
      {
        if(word.charAt(0) == 'e') //unless it is e, this avoids counting e's at the end of words
          return 0;
        else
          return 1;
      }

      int firstConsonant = 1; //the index of the first consonant
      {
        while (isVowel(word.charAt(firstConsonant)))
        {
          if(firstConsonant == word.length()-1) //if we've reached the end of the word, i.e. there were no consonants
            return 1;
          else
            firstConsonant++;
        }
      }
      return 1 + countSyllablesRecursive(word.substring(firstConsonant));
    }
    else
    {
      return countSyllablesRecursive(word.substring(1)); //count the syllables in all the characters except the first
    }
  }

  private static Boolean isVowel(char c)
  {
    switch (Character.toLowerCase(c))
    {
            case 'a':
            case 'e':
            case 'i':
            case 'o':
            case 'u':
            case 'y':
                return true;
            default:
              return false;
    }
  }

  private static Set<String> getEasyWords()
  {
    Set<String> easyWords = new HashSet<String>();
    try
    {
      Scanner scan = new Scanner(new File("/pub/pounds/CSC330/dalechall/wordlist1995.txt"));
      while (scan.hasNext())
      {
        easyWords.add(scan.next().toLowerCase());
      }
      scan.close();
    }
    catch (FileNotFoundException e)
    {
      System.err.println("Error: Could not find list of easy words for Dale-Chall.");
      System.exit(1);
    }
    return easyWords;
  }

  private static int countDifficultWords(ArrayList<String> words)
  {
    Set<String> easyWords = getEasyWords();
    int count = 0;
    for(String word : words)
    {
      if(!easyWords.contains(stripWord(word.toLowerCase())))
        count++;
    }
    return count;
  }

  private static String stripWord(String word)
  {
    //remove punctuation from beginning
    if(!Character.isLetter(word.charAt(0)))
    {
      word = word.substring(1);
    }
    //remove punctuation from end
    int lastIndex = word.length()-1;
    if(!Character.isLetter(word.charAt(lastIndex)))
    {
      word = word.substring(0, lastIndex);
    }
    return word;
  }
}

File: python/flesch.py
#!/usr/bin/python3
import sys

def isWord(word):
    for c in word:
        if c.isalpha():
            return True
    return False

def getInput():
    try:
        words = []
        with open("/pub/pounds/CSC330/translations/" + sys.argv[1], encoding="utf8", errors='ignore') as file:
            for line in file:
                for token in line.split():
                    if isWord(token):
                        words.append(token)
        return words
    except (FileNotFoundError):
        print("Error: Specified file does not exist.")
        sys.exit()

def stripWord(word):
    if not word[0].isalpha():
      word = word[1:]

    #remove punctuation from end
    if not word[-1].isalpha():
      word = word[:-1]

    return word


def countSyllablesRecursive(word):
    if len(word) == 0: #empty string base case
        return 0

    vowels = ['a', 'e', 'i', 'o', 'u', 'y']

    if word[0].lower() in vowels:

        if len(word) == 1: #if the word is only 1 character, then it is only 1 syllable
            if word == 'e': #unless it is e, this avoids counting e's at the end of words
                return 0
            else:
                return 1


        firstConsonant = 1 #the index of the first consonant

        while word[firstConsonant].lower() in vowels:
            if firstConsonant == len(word)-1: #if we've reached the end of the word, i.e. there were no consonants
                return 1
            else:
                firstConsonant += 1

        return 1 + countSyllablesRecursive(word[firstConsonant:])

    else:
        return countSyllablesRecursive(word[1:]) #count the syllables in all the characters except the first

def countSyllables(word):
    word = stripWord(word)

    numSyllables = countSyllablesRecursive(word)

    if numSyllables == 0:
      numSyllables = 1 #no word can have 0 syllables

    return numSyllables

def countTotalSyllables(words):
    count = 0
    for word in words:
      count += countSyllables(word)
    return count

def countSentences(words):
    punctuation = ['.', ':', ';', '?', '!']
    count = 0
    for word in words:
        if word[-1] in punctuation:
            count += 1
    return count

def getEasyWords():
    try:
        easyWords = set()
        with open('/pub/pounds/CSC330/dalechall/wordlist1995.txt','r') as file:
            for line in file:
                for token in line.split():
                    easyWords.add(token.lower())
        return easyWords;
    except (FileNotFoundError):
        print("Error: Dale Chall word list file does not exist.")
        sys.exit()

def countDifficultWords(words):
    easyWords = getEasyWords()
    count = 0
    for word in words:
      if stripWord(word.lower()) not in easyWords:
        count += 1
    return count

if len(sys.argv) != 2:
      print("Error: There must be exactly one command line argument.")
      sys.exit()

words = getInput()
numWords = len(words)
numSyllables = countTotalSyllables(words)
numSentences = countSentences(words)
numDifficultWords = countDifficultWords(words)

#for debugging
# print("numWords =", numWords)
# print("numSyllables =", numSyllables)
# print("numSentences =", numSentences)
# print("numDifficultWords =", numDifficultWords)

#Calculate Flesch Readability Index
alpha = numSyllables / numWords;
beta = numWords / numSentences;

fleschIndex = round(206.835 - alpha * 84.6 - beta * 1.015)

#Calculate Flesch-Kincaid Grade Level Index
fleschKincaidIndex = round(alpha * 11.8 + beta * 0.39 - 15.59, 1)

#Calculate Dale-Chall Readability Score
alpha = numDifficultWords / numWords;
dalechall = alpha * 100 * 0.1579 + beta * 0.0496;
if alpha > 0.05:
    dalechall += 3.6365
dalechall = round(dalechall, 1) #round to one decimal point

print("Flesch Readability Index =", fleschIndex);
print("Flesch-Kincaid Grade Level Index =", fleschKincaidIndex);
print("Dale-Chall Readability Score =", dalechall);
